scipy/fft/_helper.py

```diff
@@ -1,101 +1,70 @@
+from . import _helper_np as nphelper
 
 
 def next_fast_len(target, real=False):
-    """Find the next fast size of input data to ``fft``, for zero-padding, etc.
-
-    SciPy's FFT algorithms gain their speed by a recursive divide and conquer
-    strategy. This relies on efficient functions for small prime factors of the
-    input length. Thus, the transforms are fastest when using composites of the
-    prime factors handled by the fft implementation. If there are efficient
-    functions for all radices <= `n`, then the result will be a number `x`
-    >= ``target`` with only prime factors < `n`. (Also known as `n`-smooth
-    numbers)
-
-    Parameters
-    ----------
-    target : int
-        Length to start searching from. Must be a positive integer.
-    real : bool, optional
-        True if the FFT involves real input or output (e.g., `rfft` or `hfft`
-        but not `fft`). Defaults to False.
-
-    Returns
-    -------
-    out : int
-        The smallest fast length greater than or equal to ``target``.
-
-    Notes
-    -----
-    The result of this function may change in future as performance
-    considerations change, for example, if new prime factors are added.
-
-    Calling `fft` or `ifft` with real input data performs an ``'R2C'``
-    transform internally.
-
-    Examples
-    --------
-    On a particular machine, an FFT of prime length takes 11.4 ms:
-
-    >>> from scipy import fft
-    >>> import numpy as np
-    >>> rng = np.random.default_rng()
-    >>> min_len = 93059  # prime length is worst case for speed
-    >>> a = rng.standard_normal(min_len)
-    >>> b = fft.fft(a)
-
-    Zero-padding to the next regular length reduces computation time to
-    1.6 ms, a speedup of 7.3 times:
-
-    >>> fft.next_fast_len(min_len, real=True)
-    93312
-    >>> b = fft.fft(a, 93312)
-
-    Rounding up to the next power of 2 is not optimal, taking 3.0 ms to
-    compute; 1.9 times longer than the size given by ``next_fast_len``:
+    """
+    See the documentation in _helper_np.py.
+    """
+    return nphelper.next_fast_len(target, real=real)
 
-    >>> b = fft.fft(a, 131072)
 
+def _init_nd_shape_and_axes(x, shape, axes):
+    """
+    See the documentation in _helper_np.py.
     """
-    pass
+    xp = array_namespace(x)
+    x = np.asarray(x)
+    y = nphelper._init_nd_shape_and_axes(x, shape, axes)
+    return xp.asarray(y)
 
 
-# Directly wrap the c-function good_size but take the docstring etc., from the
-# next_fast_len function above
-next_fast_len = update_wrapper(lru_cache(_helper.good_size), next_fast_len)
-next_fast_len.__wrapped__ = _helper.good_size
+def fftfreq(n, d=1.0, *, xp=np, device=None):
+    """
+    Implements the array API specification of fftfreq.
+    """
+    # device is currently not supported in numpy, cupy or array-api-compat
+    if device is not None:
+        raise ValueError(
+            "Providing the 'device' parameter is not yet supported.")
```

Comment:
Is this a mistake? I'm not sure. Shouldn't we just forward the device parameter to the xp fftfreq? But if device is not None and xp == np, it should certainly raise an error.

Reply:
Forwarding the device parameter to `xp.fft.fftfreq` would work for _some_ libraries right now e.g. PyTorch. However, it won't work with CuPy, as - like NumPy - there is no `device` parameter currently. We could definitely just pass to `xp` and let CuPy throw an exception, but my reasoning for not doing this yet is to not 'jump ahead' of array-api-compat. Since array-api-compat already has helpers like `to_device`, I think that this functionality will be added once they get around to implementing `fft` (they already have `linalg`).

In fact, once array-api-compat has implemented `fft`, I think that there will be no issue with `xp=np` and `device` being not None, since we can leave array-api-compat to raise the exceptions:
```
if _is_numpy_array(x):
        if stream is not None:
            raise ValueError("The stream argument to to_device() is not supported")
        if device == 'cpu':
            return x
        raise ValueError(f"Unsupported device {device!r}")
```
(excerpt from `to_device` in array-api-compat)

Reply:
Indeed, you are totally correct. CuPy doesn't even support the `device` keyword in `fftfreq` at the moment.

I think we should do what you said, and pass it in anyway and let CuPy throw the exception. It also means we are doing the "right thing", and signals to the array libraries what exactly they need to support.

Reply:
Yes if we add parameters to comply with the standard, it's out of our hands. We only need to deal with SciPy handling it or not. e.g. if we go to compiled code instead.

Reply:
Sorry I am not following. If users pass in a np array, then the output is a np array no? Why a warning?

Reply:
No array is passed in! This is the problem and the whole need for the `xp` parameter.

Reply:
Ah yes I see. Sorry I was not paying enough attention to the surrounding code.

TBH in such cases I would not touch the functions if we don't do anything than returning the NumPy one. In the end this is something we would want to remove from SciPy. I would leave this out of this work and make a note in the PR description. People working on the module would have an opinion and I would not want you to do too much work for nothing.

Reply:
I saw this the other day saying why we would want to leave this in the SciPy API. Definitely something to think about but maybe to leave as a comment for a follow-up PR.
https://github.com/scipy/scipy/issues/13887#issuecomment-821987280

Reply:
Should I change it to just always returning the NumPy result for now? If so, then again I think it would be useful to provide a warning since this would be unexpected for someone using CuPy.

Reply:
Good find 👍 The discussion is still up so for now we have +2 and -1. We can link to that in the description of the PR.

Reply:
I would be inclined to agree with the -1 since the NumPy docs say 

> The SciPy module [scipy.fft](https://docs.scipy.org/doc/scipy/reference/fft.html#module-scipy.fft) is a more comprehensive superset of numpy.fft, which includes only a basic set of routines.

It seems to me that breaking this superset relationship might cause more issues than it is worth, but that's a decision for you guys to make!

Reply:
I'll resolve this conversation once we've decided what to do with these functions.